### PR TITLE
errdefs: remove dependency on grpc, containerd, and distribution

### PR DESCRIPTION
**Pull request by thaJeztah**
relates to https://github.com/moby/moby/pull/42623

The errdefs package is used both to generate errors, but also has utilities
to convert errors originating from subsystems into a suitable HTTP error.

The API should always return a suitable HTTP error, and Conversion of errors
originating from subsystems will (generally) only be needed in the API server
(not the client).

This patch prevents the client from depending on grpc, containerd and distribution,
by moving the code to convert those errors to a separate "adapters" package.

The api/server/httputils.GetHTTPErrorStatusCode was un-deprecated, and modified
to include the default set of adapters.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: errdefs/http_helpers.go [hunk 1]
```diff
@@ -4,21 +4,30 @@ import (
 	"fmt"
 	"net/http"
 
-	containerderrors "github.com/containerd/containerd/errdefs"
-	"github.com/docker/distribution/registry/api/errcode"
 	"github.com/sirupsen/logrus"
```
Comment (thaJeztah) [inline errdefs/http_helpers.go:7]: We should get rid of logrus here

**Comment by thaJeztah** [comment]
@dims @cpuguy83 not sure if this is the most suitable approach; let me know what you think

**Comment by dims** [comment]
@thaJeztah as long as `errdefs` are not transitively needed from moby/moby/client, we are ok for now. this looks like too much of a change  for sure. Let's hold this until we really need this, i'll POC a cadvisor change once https://github.com/moby/moby/pull/42623 lands and will let you know.

**Comment by dims** [comment]
@thaJeztah bad news ...

```
go: finding module for package github.com/containerd/containerd/errdefs
github.com/google/cadvisor/container/docker imports
	github.com/docker/docker/client imports
	github.com/docker/docker/errdefs imports
	github.com/containerd/containerd/errdefs
```

**Comment by dims** [comment]
@thaJeztah one more

```
github.com/google/cadvisor/container/docker imports
	github.com/docker/docker/client tested by
	github.com/docker/docker/client.test imports
	github.com/docker/docker/api/server/httputils imports
	github.com/docker/docker/errdefs/adapter imports
	github.com/containerd/containerd/errdefs
```

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
Having anything other than the base definitions in errdefs seems wrong.

I don't remember the change that moved stuff around, but none of this seems correct and a generic `Adapter` seems like too much abstraction.

In client/ we should only be outputting for the most part, errors that implement something in `errdefs` and we should generally only be consuming errors that come across the API.

In api/ we should only be receiving errors that implement something from errdefs and outputing the relavent status code + messages.

**Comment by thaJeztah** [comment]
> I don't remember the change that moved stuff around, but none of this seems correct and a generic Adapter seems like too much abstraction.

I think that may have been my doing. We had code to convert from/to errdefs and to/from HTTP status code in various locations, and (from that perspective) it seemed to make sense to have the errdefs package be the canonical place to provide conversions in both directions.

> In client/ we should only be outputting for the most part, errors that implement something in errdefs and we should generally only be consuming errors that come across the API.

So currently the client uses errdefs to convert HTTP status codes returned by the API back to errdefs errors (added in https://github.com/moby/moby/pull/38689). I think that makes sense (I recall string matching in various code using the client because it returned no information whatsoever).

> In api/ we should only be receiving errors that implement something from errdefs and outputing the relavent status code + messages.

So conversion from other subsystems was added over time;

- gRPC errors conversion related to SwarmKit; https://github.com/moby/moby/pull/32122
- conversion for "distribution", to prevent it returning a 500 error https://github.com/moby/moby/pull/37534
- same for "containerd" errors returning a 500 https://github.com/moby/moby/pull/39527

For those above, perhaps they should live in a client specific to each (e.g. `libcontainerd/client` for containerd ?) and convert to  our `errdef` errors?

**Comment by thaJeztah** [comment]
I had a draft alternative to this one, which does not include the `Adapter` interface, and just moves everything together into a package that's only used by the API server; I just pushed it as an alternative PR; https://github.com/moby/moby/pull/43399 - hopefully that's able to unblock this.

**Comment by thaJeztah** [comment]
Replaced by https://github.com/moby/moby/pull/43399, which was just merged 👍

**Comment by dims** [comment]
thanks a ton @thaJeztah

**Pull request closed without merging**